### PR TITLE
Flash shell migrate commands to new shell

### LINKED ISSUE #8883: Adapt samples/drivers/flash_shell to new shell subsystem
Sample is using shell subsystem thus needs to be ported once new shell is ready.

### PR DESCRIPTION
**Pull request by jakub-uC**
Migrating flash example to new shell.

Fixes #8883

### CODE REVIEW

File: samples/drivers/flash_shell/src/main.c [hunk 1]
```diff
@@ -159,200 +160,210 @@ static int do_read(off_t offset, size_t len)
...
 	if (ret) {
-		printk("flash_write_protection_set: %d (can't disable)\n", ret);
+		PR_ERROR(shell, "flash_write_protection_set: %d (can't"
+				" disable)\r\n", ret);
 		return ret;
 	}
 	ret = flash_erase(flash_device, offset, size);
 	if (ret) {
-		printk("flash_erase: %d\n", ret);
+		PR_SHELL(shell, "flash_erase: %d\r\n", ret);
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: ret != 0 indicates error. this should be PR_ERROR. Message should be something like "flash_erase failed (err:%d)."

File: samples/drivers/flash_shell/src/main.c [hunk 2]
```diff
@@ -159,200 +160,210 @@ static int do_read(off_t offset, size_t len)
...
 /* Erase area, handling write protection and printing on error. */
-static int do_erase(off_t offset, size_t size)
+static int do_erase(const struct shell *shell, off_t offset, size_t size)
 {
 	int ret;
 
 	ret = flash_write_protection_set(flash_device, false);
 	if (ret) {
-		printk("flash_write_protection_set: %d (can't disable)\n", ret);
+		PR_ERROR(shell, "flash_write_protection_set: %d (can't"
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: here message is misleading, better would be "Failed to disable flash protection (err: %d)."

File: samples/drivers/flash_shell/src/main.c [hunk 3]
```diff
@@ -159,200 +160,210 @@ static int do_read(off_t offset, size_t len)
...
 	ret = flash_erase(flash_device, offset, size);
 	if (ret) {
-		printk("flash_erase: %d\n", ret);
+		PR_SHELL(shell, "flash_erase: %d\r\n", ret);
 		return ret;
 	}
 	ret = flash_write_protection_set(flash_device, true);
 	if (ret) {
-		printk("flash_write_protection_set: %d (can't enable)\n", ret);
+		PR_ERROR(shell, "flash_write_protection_set: %d (can't"
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: "Failed to enable flash protection (err: %d)."

File: samples/drivers/flash_shell/src/main.c [hunk 4]
```diff
@@ -159,200 +160,210 @@ static int do_read(off_t offset, size_t len)
...
-static int do_write(off_t offset, u8_t *buf, size_t len, bool read_back)
+static int do_write(const struct shell *shell, off_t offset, u8_t *buf,
+		    size_t len, bool read_back)
 {
 	int ret;
 
 	ret = flash_write_protection_set(flash_device, false);
 	if (ret) {
-		printk("flash_write_protection_set: %d (can't disable)\n", ret);
+		PR_ERROR(shell, "flash_write_protection_set: %d (can't"
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: here, also messages could be improved.

File: samples/drivers/flash_shell/src/main.c [hunk 5]
```diff
@@ -159,200 +160,210 @@ static int do_read(off_t offset, size_t len)
...
+				" disable)\r\n", ret);
 		return ret;
 	}
 	ret = flash_write(flash_device, offset, buf, len);
 	if (ret) {
-		printk("flash_write: %d\n", ret);
+		PR_SHELL(shell, "flash_write: %d\r\n", ret);
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: PR_ERROR

File: samples/drivers/flash_shell/src/main.c [hunk 6]
```diff
@@ -361,217 +372,202 @@ static bool page_layout_cb(const struct flash_pages_info *info, void *datav)
 	}
 
 	sz = info->size;
-	printk("\tPage %u: start 0x%08x, length 0x%lx (%lu, %lu KB)\n",
+	printk("\tPage %u: start 0x%08x, length 0x%lx (%lu, %lu KB)\r\n",
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: you can extend struct page_layout_data to have shell handler and use PR_SHELL() here. Even though it is called callback, it is executed in the context of shell command so no worries about using shell_fprintf.

File: samples/drivers/flash_shell/src/main.c [hunk 7]
```diff
@@ -9,13 +9,26 @@
...
 #include <device.h>
 #include <soc.h>
 
-#define THIS_MODULE_NAME "flash"
+LOG_MODULE_REGISTER(app);
 
+SHELL_UART_DEFINE(shell_transport_uart);
+SHELL_DEFINE(uart_shell, "uart:~$ ", &shell_transport_uart, '\r', 10);
+
+#define PR_SHELL(shell, fmt, ...)				\
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: maybe that could be simple PR()

File: samples/drivers/flash_shell/src/main.c [hunk 8]
```diff
@@ -361,217 +372,202 @@ static bool page_layout_cb(const struct flash_pages_info *info, void *datav)
...
+	} else if (parse_ul(argv[1], &page) || parse_ul(argv[2], &offset) ||
+		   parse_ul(argv[3], &len)) {
 			goto bail;
 	}
 
 	ret = flash_get_page_info_by_idx(flash_device, page, &info);
 	if (ret) {
-		printk("flash_page_info_by_idx: %d\n", ret);
-		return 0;
+		PR_SHELL(shell, "flash_page_info_by_idx: %d\r\n", ret);
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: PR_ERROR and improve message.

File: samples/drivers/flash_shell/src/main.c [hunk 9]
```diff
@@ -361,217 +372,202 @@ static bool page_layout_cb(const struct flash_pages_info *info, void *datav)
...
+	} else if (parse_ul(argv[2], &num)) {
 		goto bail;
 	}
 
 	for (i = 0; i < num; i++) {
 		ret = flash_get_page_info_by_idx(flash_device, page + i, &info);
 		if (ret) {
-			printk("flash_get_page_info_by_idx: %d\n", ret);
-			return 0;
+			PR_SHELL(shell, "flash_get_page_info_by_idx:"
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: PR_ERROR and improve message

File: samples/drivers/flash_shell/src/main.c [hunk 10]
```diff
@@ -361,217 +372,202 @@ static bool page_layout_cb(const struct flash_pages_info *info, void *datav)
...
+				 i + 2, argv[i]);
 			goto bail;
 		}
 	}
 
 	ret = flash_get_page_info_by_idx(flash_device, page, &info);
 	if (ret) {
-		printk("flash_get_page_info_by_idx: %d\n", ret);
-		return 0;
+		PR_SHELL(shell, "flash_get_page_info_by_idx: %d\r\n", ret);
```
Comment (nordic-krch) [inline samples/drivers/flash_shell/src/main.c]: PR_ERROR

**Review by nordic-krch — CHANGES_REQUESTED** [review]

**Comment by nordic-krch** [comment]
Apart from review i have one suggestion. Why not move those commands out from sample to /drivers/flash folder and add kconfig option to enable shell commands for flash driver. Sample would just enable those.

**Comment by jarz-nordic** [comment]
@nordic-krch : done

**Review by nordic-krch — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**